LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/IoTHubImpl/IoTHubRegistryManager.cs

```diff
@@ -4,53 +4,63 @@
 
-        internal IoTHubRegistryManager(Func<RegistryManager> registryManagerProvider)
+        internal IoTHubRegistryManager(Func<RegistryManager> registryManagerProvider, IHttpClientFactory httpClientFactory, ILogger logger)
         {
             this.instance = registryManagerProvider() ?? throw new InvalidOperationException("RegistryManager provider provided a null RegistryManager.");
+            this.httpClientFactory = httpClientFactory;
+            this.logger = logger;
         }
 
-        public Task<Configuration> AddConfigurationAsync(Configuration configuration) => this.instance.AddConfigurationAsync(configuration);
-
-        public Task<Device> AddDeviceAsync(Device edgeGatewayDevice) => this.instance.AddDeviceAsync(edgeGatewayDevice);
+        public async Task<bool> AddDevice(IDeviceTwin twin)
+        {
+            var result = await this.instance.AddDeviceWithTwinAsync(new Device(twin?.DeviceId), twin.ToIoTHubDeviceTwin());
 
-        public Task<BulkRegistryOperationResult> AddDeviceWithTwinAsync(Device device, IDeviceTwin twin) => this.instance.AddDeviceWithTwinAsync(device, twin.ToIoTHubDeviceTwin());
+            if (result.IsSuccessful)
+                return true;
 
-        public Task<Module> AddModuleAsync(Module moduleToAdd) => this.instance.AddModuleAsync(moduleToAdd);
+            this.logger.LogWarning($"Failed to add Device with twin: \n{result.Errors}");
 
-        public Task ApplyConfigurationContentOnDeviceAsync(string deviceName, ConfigurationContent deviceConfigurationContent)
-            => this.instance.ApplyConfigurationContentOnDeviceAsync(deviceName, deviceConfigurationContent);
+            return false;
+        }
 
         public void Dispose() => this.instance?.Dispose();
 
-        public Task<Device> GetDeviceAsync(string deviceId) => this.instance.GetDeviceAsync(deviceId);
-
-        public async Task<IDeviceTwin> GetTwinAsync(string deviceId, CancellationToken cancellationToken)
-            => new IoTHubDeviceTwin(await this.instance.GetTwinAsync(deviceId, cancellationToken));
-
-        public async Task<IDeviceTwin> GetTwinAsync(string deviceName) => new IoTHubDeviceTwin(await this.instance.GetTwinAsync(deviceName));
-
-        public async Task<IDeviceTwin> UpdateTwinAsync(string deviceId, string moduleId, IDeviceTwin deviceTwin, string eTag)
-            => new IoTHubDeviceTwin(await this.instance.UpdateTwinAsync(deviceId, moduleId, deviceTwin.ToIoTHubDeviceTwin(), eTag));
+        public async Task<string> GetDevicePrimaryKeyAsync(string deviceId)
+        {
+            var device = await this.instance.GetDeviceAsync(deviceId);
 
-        public async Task<IDeviceTwin> UpdateTwinAsync(string deviceId, IDeviceTwin twin, string eTag, CancellationToken cancellationToken)
-            => new IoTHubDeviceTwin(await this.instance.UpdateTwinAsync(deviceId, twin.ToIoTHubDeviceTwin(), eTag, cancellationToken));
+            return device.Authentication.SymmetricKey.PrimaryKey;
```

Comment:
As said in a previous comment, if for some reason the device is null (not found), this might throw

Reply:
Same answer, the SDK returns an empty object...

Reply:
But I added a null conditional operator for this case